src/main/kotlin/com/zepben/evolve/cim/iec61970/base/core/NameType.kt

```diff
@@ -104,9 +116,32 @@ class NameType(val name: String) {
      *
      * @return true if a matching name was removed.
      */
-    fun removeNames(name: String?): Boolean = (namesIndex.remove(name) ?: namesMultiIndex.remove(name)) != null
+    fun removeNames(name: String?): Boolean {
+        // Calling removeName from identifiedObject will remove the name from both the identifiedObject and the nameType
+        namesIndex[name]?.let { n->
+            return n.identifiedObject.removeName(n)
+        }
+        namesMultiIndex[name]?.let { names->
+            var removed = false
+            names.toList().forEach { name->
+                removed = name.identifiedObject.removeName(name)
+            }
+            return removed
+        }
+        return false
+    }
 
     fun clearNames(): NameType {
+        // Remove name from associated identified object when it's removed from the nameType
+        namesIndex.toMap().forEach { entry->
```

Comment:
suggestion: you can make the looping over a map more readable using destructing syntax. e.g:
```
namesIndex.toMap().forEach { (_, name) ->
    name.identifiedObject.removeName(name)
}
```

you could also use access via the `values` collection. e.g:
```
namesIndex.values.toList().forEach {
    it.identifiedObject.removeName(it)
}
```

Reply:
picked the destructing syntax.